api/lql.go

```diff
@@ -18,23 +18,115 @@
 
 package api
 
+import (
+	"encoding/json"
+	"net/url"
+	"regexp"
+)
+
+const (
+	reLQL string = `(?ms)^(\w+)\(\w+\s\w+\)\s*{`
+)
+
+type LQLQuery struct {
+	ID             string `json:"LQL_ID,omitempty"`
+	StartTimeRange string `json:"START_TIME_RANGE,omitempty"`
+	EndTimeRange   string `json:"END_TIME_RANGE,omitempty"`
+	QueryText      string `json:"QUERY_TEXT"`
+	// QueryBlob is a special string that supports type conversion
+	// back and forth from LQL to JSON
+	QueryBlob string `json:"-"`
+}
+
+func (q *LQLQuery) translate() {
+	// if QueryText is not already populated
+	if q.QueryText == "" {
+		// if QueryBlob is JSON
+		var t LQLQuery
+		err := json.Unmarshal([]byte(q.QueryBlob), &t)
+
+		if err == nil {
+			q.StartTimeRange = t.StartTimeRange
+			q.EndTimeRange = t.EndTimeRange
+			q.QueryText = t.QueryText
+			// if QueryBlob is LQL
+		} else if matched, _ := regexp.MatchString(reLQL, q.QueryBlob); matched {
+			q.QueryText = q.QueryBlob
+		}
+	}
```

Comment:
nit: we like to keep the nested code as little as possible, this could be change to

```suggestion
	// only if QueryText is empty
	if q.QueryText != "" {
		return
	}
	// if QueryBlob is JSON
	var t LQLQuery
	if err := json.Unmarshal([]byte(q.QueryBlob), &t); err == nil {
		q.StartTimeRange = t.StartTimeRange
		q.EndTimeRange = t.EndTimeRange
		q.QueryText = t.QueryText
		return
	}

	// if QueryBlob is LQL
	if matched, _ := regexp.MatchString(reLQL, q.QueryBlob); matched {
		q.QueryText = q.QueryBlob
	}
```

Reply:
Fixed.